core/include/keyman/keyman_core_api.h

```diff
@@ -526,6 +558,128 @@ enum km_core_action_type {
   KM_CORE_IT_MAX_TYPE_ID
 };
 
+/*
+```
+### Actions
+This structure provides the results of processing a key event to the Platform layer and
+should be processed by the Platform layer to issue commands to the os text
+services framework to transform the text store in the Client Application, among
+other actions.
+
+This API replaces the Action items APIs, which is now deprecated and will be
+removed in the future.
+```c
+*/
+
+typedef struct {
+  // number of codepoints (not codeunits!) to delete from app context, 0+
+  int delete_back;
+
+  // null-term string of characters to insert into document
+  km_core_usv* output;
+
+  // list of options to persist, terminated with KM_CORE_OPTIONS_END
+  km_core_option_item* persist_options;
+
+  // issue a beep
+  bool do_alert;
+
+  // emit the input keystroke to the application, unmodified?
+  bool emit_keystroke;
+
+  // -1=unchanged, 0=off, 1=on
+  int new_caps_lock_state;
+} km_core_actions;
+
+/*
+```
+### `km_core_state_get_actions`
+##### Description:
+Returns a pointer to an actions object which details all the actions
+that the Platform layer must take after a keystroke. The `delete_back`
+action must be performed before the `output` action, but the other
+actions may be performed in any order.
+##### Return:
```

Comment:
I'm wondering if we can name the delete_back differently to include the two important comments about it and make it more self-documenting: 1) it is number of code points, not code units and 2) it must be executed before the output.
The name for that could be long, but it should only be consumed once in the platform code and would help to not have to search the docs to understand the proper way to apply it.

Reply:
`code_points_to_delete` ? seems like a good idea

Reply:
I'm good with `delete_back_codepoints` but I think reading the API docs is essential for using an API correctly, because you can't really embed the _reasoning_ in the name without causing additional grief. I'm not that keen on adding `_first` because the entire state machine pattern needs to be understood anyway and `_first` actually confuses things IMO. What does `_first` refer to -- is it order of operations, or is it delete from first item in buffer, or ...? We'd have to be too wordy to make it helpful.

Reply:
I think code_points_to_delete is an improvement. Or you could try code_points_to_delete_before_inserting_text, but I have a feeling you would consider that too wordy. Isn't it friendly and readable though? :)

Reply:
`code_points_to_delete_before_inserting_text` (44 chars) would probably run into some maximum symbol size issue on one of our platforms 😉 

[For example](https://stackoverflow.com/a/57440831/1836776):
> Rule 5.1 (required): Identifiers (internal and external) shall not rely on the significance of more than 31 characters